src/engine/Account-Imap.cpp

```diff
@@ -0,0 +1,942 @@
+                  tokenInfo.total_token_count);
+
+            auto itr = probability_hash.find (account_c->account_guid);
+            if (itr != probability_hash.end()) // existing ?
+            {
+                // Existing entry
+                PINFO("Existing entry for this account");
+
+                itr->second.m_product = (((double)account_c->token_count /
+                                       (double)tokenInfo.total_token_count)
+                                        * itr->second.m_product);
+
+                itr->second.m_product_difference = ((double)1 - ((double)account_c->token_count /
+                                                  (double)tokenInfo.total_token_count))
+                                                   * itr->second.m_product_difference;
+
+                PINFO("product == %f, product_difference == %f",
+                      itr->second.m_product, itr->second.m_product_difference);
+            }
+            else
+            {
+                // New entry
+                PINFO("Adding a new entry for this account");
+
+                /* set the product and product difference values */
+                double product = ((double)account_c->token_count /
+                                  (double)tokenInfo.total_token_count);
+
+                double product_difference = (double)1 - ((double)account_c->token_count /
+                                            (double)tokenInfo.total_token_count);
+
+                account_probability a_prob({product, product_difference});
+
+                probability_hash.emplace(account_c->account_guid, a_prob);
+
+                PINFO("product == %f, product_difference == %f",
+                      product, product_difference);
+            }
+        } // for all accounts in tokenInfo
+
+        // Now look for highest probability
+        for (auto &itr : probability_hash)
+        {
+            /* P(AB) = A*B / [A*B + (1-A)*(1-B)]
+             * NOTE: so we only keep track of a running product(A*B*C...)
+             * and product difference ((1-A)(1-B)...) */
+
+            int32_t probability =
+                (itr.second.m_product /
+                (itr.second.m_product + itr.second.m_product_difference))
+                 * PROBABILITY_FACTOR;
+
+            if (probability > max_probability)
+            {
+                max_probability = probability;
+                selected_account_guid = itr.first;
+            }
+        }
+
+        /* free the data in tokenInfo */
+        for (current_account_token = tokenInfo.accounts; current_account_token;
+                current_account_token = current_account_token->next)
+        {
+            /* free up each struct account_token_count we allocated */
+            delete((struct account_token_count*)current_account_token->data);
+        }
+        g_list_free (tokenInfo.accounts); /* free the accounts GList */
+    }
```

Comment:
It would be vastly better for this to go in ```~token_accounts_info()```. It would be even more vastly better to use a ```std::vector<std::shared_ptr<account_token_count>>``` instead of a GList, because that would get rid of a GList (very slow, especially if it's used without the GSlice allocator and not type safe) and the compiler would take care of all of the cleanup so you wouldn't even need ```~token_accounts_info()```. Plus you could use std::find() and range-for on the std::vector for less code and more speed.
Consider as well replacing ```struct account_token_count``` with ```std::pair<std::string, int64_t>``` which saves you the work of a constructor; just call ```std::make_pair(key, count)```. Similarly, ```struct token_accounts_info``` can be replaced with ```std::pair<std::vector<std::shared_ptr<std::pair<std::string, int64_t>>>, int64_t>```, though you'd probably want to alias them for readability:
```
using AccountTokenCount = std::pair<std::string, int64_t>;
using AccountTokenCountPtr = std::shared_ptr<AccountTokenCount>;
using TokenAccountsInfo = std::pair<std::vector<AccountTokenCountPtr>, int64_t>;
```